packages/grid/x-data-grid-premium/src/concept/DataGrid.tsx

```diff
@@ -0,0 +1,606 @@
+    const ctx = this.canvas.current!.getContext('2d')!
+
+    const w = this.canvas.current!.width
+    const h = this.canvas.current!.height
+
+    ctx.clearRect(0, 0, w, h)
+    ctx.fillStyle = 'rgba(255, 255, 255, 0.1)'
+    ctx.fillRect(0, 0, w, h)
+
+    const f = 0.1
+    const cellHeight = this.rowHeight * f * 0.15
+    const cellWidth = Math.round(this.columnWidth * f * 0.4)
+
+    ctx.lineWidth = 0.125
+
+    const doRect = (range: Range, op = ctx.strokeRect.bind(ctx)) => {
+      const x = range.minX * cellWidth
+      const y = range.minY * cellHeight
+      const w = (range.maxX + 1) * cellWidth - x - ctx.lineWidth
+      const h = (range.maxY + 1) * cellHeight - y - ctx.lineWidth
+      op(x + 0.5, y + 0.5, w, h)
+    }
+
+    const blocks = this.blockTree.all()
+    for (const block of blocks) {
+      ctx.strokeStyle = block.didRender ? 'green' : 'blue'
+      doRect(block)
+    }
+
+    ctx.lineWidth = 0.25
+    ctx.strokeStyle = 'red'
+    doRect(this.displayRange)
+
+    ctx.strokeStyle = 'yellow'
+    ctx.fillStyle = 'rgba(255, 255, 0, 0.1)'
+    doRect(this.screenRange)
+    doRect(this.screenRange, ctx.fillRect.bind(ctx))
+
+
+    this.canvasFrameId = requestAnimationFrame(this.updateCanvas)
+  }
+
+  render() {
+    return (
+      <div>
+        <div
+          className='grid'
+          style={{
+            '--row-height': `${this.rowHeight}px`,
+            '--header-height': `${this.rowHeight}px`,
+          } as React.CSSProperties}
+        >
+          <div ref={this.scrollVertical} className='fake-scroll-vertical'>
+            <div className='fake-scroll-vertical-content'></div>
+          </div>
+
+          <div ref={this.container} className='container'>
+            <div className='header'>
+              {this.columns.map((c) =>
+                <div className='column' key={c.field}>{c.headerName ?? c.field}</div>
+              )}
+            </div>
+
+            <div ref={this.content} className='content'></div>
+          </div>
+        </div>
+        <br/>
+        <canvas ref={this.canvas} width='700' height='400' />
+      </div>
+    );
+  }
+}
+
+function Block({ instance, rowIndex, columnIndex, onDidMount, onDidUnmount }: {
+  instance: DataGrid,
+  rowIndex: number;
+  columnIndex: number;
+  onDidMount?: Function;
+  onDidUnmount?: Function;
+}) {
+  React.useEffect(() => {
+    onDidMount?.()
+    return onDidUnmount as any
+  }, [])
+
+  return (
+    <React.Fragment>
+      {instance.rows.slice(rowIndex, rowIndex + BLOCK_ROW_SIZE).map((row, i) =>
+        <div className='row' key={i}>
+          {instance.columns.slice(columnIndex, columnIndex + BLOCK_COL_SIZE).map((column, j) =>
+            <div className='column' key={j}>
+              {column.valueFormatter!({ value: row[column.field] } as any)}
+            </div>
+          )}
+        </div>
+      )}
+    </React.Fragment>
+  )
+}
+
+function afterAnimationFrame(fn: Function) {
+  requestAnimationFrame(() => {
+    setImmediate(fn)
```

Comment:
The `afterAnimationFrame` function is there to allow starting work right after the browser has painted a frame.

Reply:
Aaah, so `requestAnimationFrame` executes before the paint, and then you use `setImmediate` to schedule a microtask that will be run after the paint?

Reply:
Yes. I've lifted the `setImmeditate` implementation from React's code, they use `MessageChannel` ports which seems to be the best method for an immediate callback. I think it's a task, not a microtask, possibly because a microtask would run after the RAF handlers but before the paint.